### PR TITLE
makefiles: Use simple expansion for widely used variables

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description
This PR changes some Makefile variable assignments to simple expansion, so we don't have to shell out multiple times when not required on widely used variables.

One can see all the calls to `basename`, for instance, by doing `make SHELL='/bin/sh -x' 2>&1 | less`.

### Test results 
I built **50 times** a couple of applications, for different boards, and measured with `time` each build. These are the averages:
#### samr21-xpro:`examples/gcoap`
- On master:
  - Real    3.85 s
  - User    2.95 s
  - System  0.96 s

- With this PR:
  - Real    3.02 s
  - User    2.38 s
  - System  0.66 s

**Time decreases in ~21.5%**.

#### native:`examples/gnrc_networking`
- On master:
  - Real    2.28 s
  - User    1.63 s
  - System  0.70 s

- With this PR:
  - Real    1.91 s
  - User    1.33 s
  - System  0.60 s

**Time decreases in ~16.2%**.

### Testing procedure
- Green Murdock
- You can measure time with something like:
```bash
BOARD=samr21-xpro make -C examples/gcoap clean
time -p BOARD=samr21-xpro make -C examples/gcoap all
```

### Issues/PRs references
None

### CODE REVIEW

File: Makefile.base [hunk 1]
```diff
@@ -14,7 +14,8 @@ endif
 unexport DIRS
 DIRS := $(sort $(abspath $(DIRS)))
 
-MODULE ?= $(shell basename $(CURDIR))
+_MOD := $(shell basename $(CURDIR))
+MODULE ?= $(_MOD)
```
Comment (cgundogan) [inline Makefile.base]:
```sh
make SHELL='/bin/sh -x' 2>&1 | less
```
does indeed show a huge number of superfluous shell-outs to basename. Good catch!

**Comment by cgundogan** [comment]
let's see how murdock reacts to this optimization

**Comment by kaspar030** [comment]
Nice one!
I can confirm the speed increase:

This PR:

```
[kaspar@ng riot.tmp/examples/gcoap (pr13292)]$ hyperfine -w1 -m20 "BOARD=samr21-xpro make -j8"
Benchmark #1: BOARD=samr21-xpro make -j8                                                                                         
  Time (mean ± σ):      1.032 s ±  0.028 s    [User: 1.543 s, System: 0.341 s]
  Range (min … max):    1.009 s …  1.088 s    20 runs
```

master:

```
[kaspar@ng riot.tmp/examples/gcoap ((769209351a...))]$ hyperfine -w1 -m20 "BOARD=samr21-xpro make -j8"
Benchmark #1: BOARD=samr21-xpro make -j8
  Time (mean ± σ):      1.224 s ±  0.073 s    [User: 1.984 s, System: 0.447 s]
  Range (min … max):    1.146 s …  1.395 s    20 runs
```

1.032 vs 1.224 sec, on a Thinkpad with i5-8250U, Linux, performance governor.

**Comment by cgundogan** [comment]
>Nice one!

@kaspar030 murdock does not seem to be affected by this speed increase. Any ideas why?

**Comment by fjmolinas** [comment]
Nice catch! @leandrolanzieri are these variables **allways**? if not maybe you could use `memoized`?

**Comment by kaspar030** [comment]
> @kaspar030 murdock does not seem to be affected by this speed increase. Any ideas why?

Murdock builds *massively* parrallel (e.g., N times "make -j4" for N cores), in order to keep the CPUs busy. I'd assume that the fix in this PR doesn't increase the throughput much in that setting, CPU's waiting to be fed are just scheduled to compile something else.

But it might actually do speedup some things. Murdock build times are quite variable, but if you compare the average build times of the workers "riotbuild" and "mobi9..." (both have 64 cores, the former Epycs, the latter some older Opterons IIRC),  they were significantly faster in this PR vs. the previous two successfully builds.

this PR: https://ci.riot-os.org/RIOT-OS/RIOT/13292/fdff969695bdb3ef4b67763da6766b93aa7c498e/output.html
Previous: https://ci.riot-os.org/RIOT-OS/RIOT/13290/5391e50bb7721895d3406ed1dde7352113b74190/output.html
Another: https://ci.riot-os.org/RIOT-OS/RIOT/12969/c4e0ed4b7991683f6dfd25fd2905dcc43e0d371e/output.html

(scroll all the way down to see the worker stats)

**Comment by kaspar030** [comment]
> @leandrolanzieri are these variables **allways**? if not maybe you could use `memoized`?

That sounds interesting, but I have no clue what it means. Do you have a pointer?

**Comment by leandrolanzieri** [comment]
> Nice catch! @leandrolanzieri are these variables **allways**? 

I think in the case of `MODULE` yes, as it is used for building the targets in `Makfile.base`.

> if not maybe you could use `memoized`?

I can measure using  `memoized` for the other variables.

**Comment by leandrolanzieri** [comment]
> > @leandrolanzieri are these variables **allways**? if not maybe you could use `memoized`?
> 
> That sounds interesting, but I have no clue what it means. Do you have a pointer?

`memoized` is defined in `makefiles/utils/variables.mk`.

**Comment by leandrolanzieri** [comment]
> Nice catch! @leandrolanzieri are these variables **allways** used?

Also, I see now `OBJCOPY` is evaluated right after so it's expanded.

**Comment by cgundogan** [comment]
Seems to be fine and murdock also agrees. GO!

**Pull request merged**